### DISCUSSION TITLE
Sverchok and FreeCAD 1.0 incompatibility

### DISCUSSION THREAD

**Discussion by rendetto** [nortikin/sverchok#5174] [General]
For those who struggle...

FreeCAD 1.0 official was released at the end of November. The version of Python it uses seems aligned to Blender's - it's 3.11.x , but actually they differ. 
Blender4.x (even the experimental 4.4.0alpha) uses Python 3.11.9 while the newest FreeCAD 1.0 series uses 3.11.10. The result is fail loading the Part module:

`ImportError: DLL load failed while importing Part: The specified procedure could not be found.`

The last release that was working for me was 1.0RC2 (https://github.com/FreeCAD/FreeCAD-Bundle/releases/tag/1.0rc2) which still uses Python 3.11.9

**Comment by rendetto** [comment]
Still wandering into this compatibility issue. 

There is an article form Yorik van Havre with a sample code that embeds FreeCAD as a library and uses the Part module inside Blender https://wiki.freecad.org/Embedding_FreeCAD
It was a bit old so it needed an update to work with the Blender4.3.2 I'm testing with. 
It's a script that can import solid objects from .FCStd file converting those to Blender meshes through the import procedure. 
Here is the updated code:
```
import bpy
import sys

# Path to FreeCAD library
FREECADPATH = 'D:\\FreeCAD_weekly-builds-39747-conda-Windows-x86_64-py311\\bin'
sys.path.append(FREECADPATH)

def import_fcstd(filename):
    try:
        import FreeCAD
        import Part
    except ImportError:
        print("Error: FreeCAD library not found. Check the FREECADPATH variable.")
        return

    try:
        doc = FreeCAD.open(filename)
    except Exception as e:
        print(f"Error opening FreeCAD file: {e}")
        return

    objects = doc.Objects
    for ob in objects:
        if ob.TypeId.startswith('Part::'):
            shape = ob.Shape
            if shape.Faces:
                mesh = bpy.data.meshes.new(ob.Name)
                obj = bpy.data.objects.new(ob.Name, mesh)
                bpy.context.collection.objects.link(obj)

                verts = []
                faces = []
                tessellation = shape.tessellate(1)
                verts = [(v.x, v.y, v.z) for v in tessellation[0]]
                faces = [tuple(f) for f in tessellation[1]]

                mesh.from_pydata(verts, [], faces)
                mesh.update()

    FreeCAD.closeDocument(doc.Name)

def main():
    bpy.ops.wm.call_menu(name="IMPORT_FCSTD")

class ImportFCSTDOperator(bpy.types.Operator):
    """Import FCSTD File"""
    bl_idname = "import_scene.fcstd"
    bl_label = "Import FCSTD"
    bl_options = {'REGISTER', 'UNDO'}

    filepath: bpy.props.StringProperty(subtype="FILE_PATH")

    def execute(self, context):
        import_fcstd(self.filepath)
        return {'FINISHED'}

    def invoke(self, context, event):
        context.window_manager.fileselect_add(self)
        return {'RUNNING_MODAL'}

def menu_func_import(self, context):
    self.layout.operator(ImportFCSTDOperator.bl_idname, text="FreeCAD (.fcstd)")

def register():
    bpy.utils.register_class(ImportFCSTDOperator)
    bpy.types.TOPBAR_MT_file_import.append(menu_func_import)

def unregister():
    bpy.utils.unregister_class(ImportFCSTDOperator)
    bpy.types.TOPBAR_MT_file_import.remove(menu_func_import)

if __name__ == "__main__":
    register()

```
You need to specify FREECADPATH to match your FreeCAD bin folder. Make sure that you use `\\` instead of a single `\` under Windows.
When the script is executed from Blender's Text editor an import menu is present under File>Import as FreeCAD(.fcstd).

**Comment by rendetto** [comment]
A small update... I messed up the FreeCAD Python versions combined with not properly updating the FREECADPATH so it appeared to me at first that newer versions of FreeCAD (later than 1.0RC2) were actually compatible. 
In reality, after I rechecked, the Python versions difference is still the possible cause of issues. 
The import script above is throwing an error and not working with the latest FreeCAD.

**Comment by rendetto** [comment]
As an addition to the above and especially for getting FreeCAD library to work with Sverchok, or at least to properly check if the specific FreeCAD/Python version works with Blender:

1. Make sure Blender and FreeCAD use the same versions of Python. 
Sometimes is not enough the major.minor numbers to be the same. As an example I checked:
Blender3.5.1(Python3.10.9)/FreeCAD0.20.0(Python3.10.10) - works
Blender3.5.1(Python3.10.9)/FreeCAD0.22.0(Python3.10.13) - works
Blender3.5.1(Python3.10.9)/FreeCAD1.0.0dev(Python3.10.16) - does not work (FreeCAD as a custom build conda package in this example)
Blender4.3.2(Python3.11.9)/FreeCAD1.1.0dev_weekly38728(Python3.11.09) - works
Blender4.3.2(Python3.11.9)/FreeCAD1.0.0RC2(Python3.11.9) - works
Blender4.3.2(Python3.11.9)/FreeCAD1.0.0RC3(Python3.11.10) - does not work
Blender4.3.2(Python3.11.9)/FreeCAD1.0.0official(Python3.11.10) - does not work
Blender4.3.2(Python3.11.9)/FreeCAD1.0.0dev_weekly39747(Python3.11.11)  - does not work

2. Make sure that you've set the path to FreeCAD bin folder from **freecad_path.pth** that needs to be located in your Blender Python "site-packages" folder e.g. `D:\blender-4.2.2-windows-x64\4.3\python\lib\site-packages` (see https://github.com/nortikin/sverchok/discussions/5180 ). This file usually should be automatically created when you point to FreeCAD location from Sverchok UI but it's not uncommon that you need to set this manually and I recommend the manual method.

3. After any change in scripts configuration or code it's a good habit to delete all `__pycache__` folders before starting Blender. Those are located in your Blender installation folder and also in the user specific Blender folders (under Windows those are usually located in your profile folder e.g. `C:\Users\...\AppData\Roaming\Blender Foundation\Blender\4.3`). For example I had to remove all the pycache from Blender3.5.1 in order to update the FreeCAD path installation, because something in pycache remembered an old working configuration and pointed to it despite I changed the path in **freecad_path.pth**.

**Comment by rendetto** [comment]
I'm puzzled. I created a conda package of the latest FreeCAD with the following command 
`conda create -p fc_env_py3119 freecad python=3.11.9 --copy -y -c conda-forge`
which created a FreeCAD1.0.0 that uses internally Python3.11.9
Unexpectedly it turns out being incompatible with Blender/Sverchok.

**Comment by adiazubierna** [comment]
I'm trying to make it work too. Sadly I am not an expert so I can't hel you too much. But I spent the whole day trying to make it work and I failied... Any extra test you may need, I will happy to help. Also if you can solve it, I would appreciatte if you can share the solution.

**Comment by rendetto** [comment]
```
import bpy
import sys

def check_modules(library_path=None):
    # If a path is provided, add it to sys.path
    if library_path:
        sys.path.append(library_path)
    
    try:
        import FreeCAD
        import Part
        print("FreeCAD and Part modules loaded successfully!")
    except ImportError as e:
        print(f"Error loading modules: {e}")

# Define the path to the FreeCAD library here (e.g., "C:/Program Files/FreeCAD 0.19/lib")
freecad_library_path = "D:\\FreeCAD_1.0.0RC2-conda-Windows-x86_64-py311\\bin"

# Run the check when the script is executed
check_modules(freecad_library_path)
```
This simple script checks if FreeCAD library and Part module can be loaded at all independently from Sverchok. Run it from Blender's Text editor. It's compatible with Blender4.3.2 (I did not tested it on other releases). You need to restart Blender when you change the freecad_library_path variable in order to refresh the path data.
It helps testing other FreeCAD releases for compatibility while Sverchok is running a compatible FreeCAD.

**Comment by portnov** [comment]
https://github.com/nortikin/sverchok/tree/master/scripts these examples may appear helpful for such testing scripts.

**Comment by adiazubierna** [comment]
Hello,

I was able to install Blender 4.3.2 with FreeCAD 1.0.0RC2 instead of the last stable version based on @rendetto coment on other discussion (thank you so much), but sadly when I am trying some nodes it gives me error... Any ideas how to solve it?


![Captura de pantalla 2025-01-04 202946](https://github.com/user-attachments/assets/9f2f2d17-f94f-45e4-9d93-2d2c7b3b8702)

Also the error is the following one:

`
Python: Traceback (most recent call last):
  File "C:\Users\Alberto\AppData\Roaming\Blender Foundation\Blender\4.3\scripts\addons\sverchok-master\nodes\exchange\approx_subd_to_nurbs.py", line 34, in execute
    node.Approximate(node)
  File "C:\Users\Alberto\AppData\Roaming\Blender Foundation\Blender\4.3\scripts\addons\sverchok-master\nodes\exchange\approx_subd_to_nurbs.py", line 67, in Approximate
    S = ApproxSubdToNurbs( node.inputs['Subd Obj'].sv_get()[0] )
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Alberto\AppData\Roaming\Blender Foundation\Blender\4.3\scripts\addons\sverchok-master\nodes\exchange\approx_subd_to_nurbs.py", line 98, in ApproxSubdToNurbs
    if obj.modifiers[0].levels <= 1:
       ~~~~~~~~~~~~~^^^
IndexError: bpy_prop_collection[index]: index 0 out of range, size 0

`

I am trying to set a workflow to transform from STL file to solid, and I will document everything but now I am stucked.

Thank you,

**Reply by rendetto** [reply]
This node is not yet documented. Generally you need to apply Subdivision Surface modifier from Blender onto the input object in order for the node to work. 
I recommend moving this request to a different discussion, since it hijacks the topic it's discussed here. You can just delete your comment and create another topic, so we can keep things easier to track. Thanks.

**Comment by rendetto** [comment]
Good news! Blender 4.4.0 comes aligned with the current weekly FreeCAD1.0 builds. Both use Python 3.11.11 now.
"Solids" and other dependent nodes are working again.

**Comment by chennes** [comment]
FreeCAD 1.1 will be released Soon™ -- it's currently being built with 3.11.13. Have the dev weeklies been tested with Sverchok?

**Comment by rendetto** [comment]
I can confirm that latest FreeCAD_weekly-2025.09.24-Windows-x86_64-py311(Python 3.11.13) is still compatible with Blender 4.4.0(Python 3.11.11).

**Comment by RockVicedomini** [comment]
The latest FREECAD 1.1.rc2 (.appimage on Linux) has python 3.11.14. 
I have been able to setup Sverchok nodes in Blender 5.0.1, but the node read FCSTD crashes (Blender executable is provided with Python  3.11.13. 
Can this issue being investigated before the release of the official FREECAD 1.1?

**Comment by rendetto** [comment]
I didn't checked Blender 5.0.1 with FreeCAD 1.1.rc2/3 (I'm still on Blender 5.0.0). But there could be compatible versions still. Just test and find what's working for you.
The node "Read FCStd" can crash not only from python version incompatibility but with unsupported data. You just need to test what works and what not. If you can do any other Solid operation, this means compatibility is still there and if something particular crashes this is just due to the experimental nature of the node you are using - so you need to approach it as a bug and make a report.

**What is clear now is that the official Blender 5.1.0 is on Python 3.13.9 and official FreeCAD 1.1.0 is on Python 3.11.14 - apparently no compatibility for the near future.**

A positive thinking about this situation is you rarely need to use the last version of this or that unless for very specific purpose. The important goal is usually to get the job done, so versions are not so important.

I usually have several Blender and FreeCAD versions on disposal, because if you hit a bug or regression problem making a good report is not possible unless you debug backwards to the last working state and this usually means you need to fire up several old versions in combination. It's frustrating but that's the cost of progress.

**Comment by nortikin** [comment]
4.2 last working version of blender for linux and windows also. with current sverchok.
But i cannot understand what rfeecad changed, that all dropped

**Reply by nortikin** [reply]
even if freecad and blender python match, different compilation toolset breaks down. Seems like it is freecad issue itself, maybe we can utilise https://pyocct.readthedocs.io/en/latest/

**Reply by rendetto** [reply]
PyOCCT development seems a bit stall. PythonOCC  and CADQuery-OCP are viable options but I tend towards CADQuery-OCP from what I'm reading here and there.

**Comment by dongmingchao** [comment]
I just make this work, blender 5.0.1 + FreeCAD 1.1 on Win11. After so much failure, I just **use binary file from Offical Releases, NOT conda**
<img width="712" height="299" alt="image" src="https://github.com/user-attachments/assets/9d6149a9-1441-4fc4-b029-fdd389cb0dce" />

.pth file
<img width="1133" height="186" alt="image" src="https://github.com/user-attachments/assets/755cfbcf-43d3-41dd-b942-768cb04e191c" />

I am trying to `Read FCStd`, but just a little parts could be seen, like Cube/Cone, Fusion(by Part workbench, not PartDesign), Compose.
The biggest problem is `Solid to Mesh` node broken, I can not do render work.
<img width="1353" height="995" alt="image" src="https://github.com/user-attachments/assets/e2582230-24c1-43d0-822d-6b8d1d5c2464" />

Anyway, Thanks you guys, it's a very nice workflow.

**Comment by sp93** [comment]
Sverchok with Blender 5.0.1 works with FreeCAD 1.1.1 and 1.0.2
All three have python 3.11

I tried [dongmingchao](https://github.com/dongmingchao)'s suggested method but Sverchok wasn't registering FreeCAD's path through the preferences which eventually ends up as a .pth file in Blender's site-packages.

I created freecad_path.pth inside Blender Foundation\Blender 5.0\5.0\python\lib\site-packages and straight up pasted my freecad bin directory path in plain text - it worked. I tried this for both, FreeCAD 1.1.1 and 1.0.2, they worked regardless of whether my freecad was from a binary or a conda distribution.
